src/app/subjects/services/subject-table-body.service.ts

```diff
@@ -0,0 +1,152 @@
+    for (let i: number = 0; i < body.length; i++) {
+      if (+body[i].id === mark.studentId) {
+        return body[i];
+      }
+    }
+
+    return null;
+  }
+
+  private getStudentByMarkId(students: Array<Student>, mark: Mark): TNullable<Student> {
+    for (let i: number = 0; i < students.length; i++) {
+      if (students[i].id === mark.studentId) {
+        return students[i];
+      }
+    }
+
+    return null;
+  }
+
+  private computeAverageMarkByField(field: ICell<string>): number {
+    let count: number = 0;
+    const sum: number = Object.keys(field).reduce((acc, key) => {
+      return Number.isNaN(+key) ? acc : (count++, acc + +field[key]);
+    }, 0);
+
+    return count === 0 ? -1 : sum / count;
+  }
+
+  public createBody(marks: Array<Mark>, students: Array<Student>): Array<ICell<string>> {
+    const body: Array<ICell<string>> = [];
+
+    marks.forEach((mark) => {
+      let field: TNullable<ICell<string>> = this.findBodyByMarkId(body, mark);
+      let isCreated: boolean = false;
+
+      if (field === null) {
+        const student: TNullable<Student> = this.getStudentByMarkId(students, mark);
+        if (student === null) {
+          return;
+        }
+
+        field = this.createBodyField(student);
+        isCreated = true;
+      }
+
+      field[mark.date] = `${mark.value}`;
+      if (isCreated) {
+        body.push(field);
+      }
+    });
+
+    return body;
+  }
+
+  public updateMarksByDate(body: Array<ICell<string>>, changes: DateChanges): Array<ICell<string>> {
+    const { current, previously: prev } = changes;
+
+    return body.map((field) => {
+      if (!field[prev]) {
+        return field;
+      }
+
+      const item: ICell<string> = Object.assign({}, field);
+      item[current] = item[prev];
+      delete item[prev];
+      return item;
+    });
+  }
+
+  public updateMark(
+    body: Array<ICell<string>>,
+    change: IChangeField<number>,
+  ): Array<ICell<string>> {
+    const { value: mark, column: date, row: rowIndex } = change;
+
+    return body.map((item, index) => {
+      if (index !== rowIndex) {
+        return item;
+      }
+
+      if (mark !== -1) {
+        item[date] = `${mark}`;
+      } else {
+        delete item[date];
+      }
+
+      return item;
+    });
+  }
+
+  public deleteMarksByDate(milliseconds: number, body: Array<ICell<string>>): Array<ICell<string>> {
+    const newBody: Array<ICell<string>> = body.map((item) => {
+      const newItem: ICell<string> = Object.assign({}, item);
+      delete newItem[milliseconds];
+      return newItem;
+    });
+
+    this.computeAverageMarkByBody(newBody);
+    return newBody;
+  }
+
+  public computeAverageMarkByBody(body: Array<ICell<string>>): void {
+    body.forEach((item) => {
+      const averageMark: number = this.computeAverageMarkByField(item);
+      if (averageMark === -1) {
+        return (item['average mark'] = '');
+      }
+
+      item['average mark'] = `${averageMark}`;
```

Comment:
Cool if this method would be pure, that means it will only return averageMark, and where this function is called this computed value would be set to proper 'average mark' column